### PR TITLE
Break apart different BrowserFragment versions

### PR DESCRIPTION
**Pull request by NotWoods**
@csadilek and I discussed this morning how to separate the different entry points for `BrowserFragment` (normal browsing, custom tabs, PWAs). We ended up settling on keeping the flat list of features structure, with 3 variant classes that append a few additional features depending on what entry point they're used for.

Whiteboard showing separation of features

This PR also includes two other changes:
- Pass `GoBackUseCase` directly to `SessionFeature`
- Fix `BrowserMenu` so that "Add to homescreen" doesn't show up in custom tabs.

### CODE REVIEW

File: samples/browser/src/main/java/org/mozilla/samples/browser/CustomTabBrowserFragment.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                components.menuBuilder,
+                closeListener = { activity?.finish() }),
+            owner = this,
+            view = layout)
+
+        return layout
+    }
+
+    override fun onBackPressed(): Boolean =
+        super.onBackPressed() || customTabsToolbarFeature.onBackPressed()
```
Comment (jonalmeida) [inline samples/browser/src/main/java/org/mozilla/samples/browser/CustomTabBrowserFragment.kt]: You probably want to evaluate the feature's `onBackPressed` first before super. Same goes for the `WebAppBroserFragment.kt`.

Comment (NotWoods) [reply samples/browser/src/main/java/org/mozilla/samples/browser/CustomTabBrowserFragment.kt]: This matches the original behaviour where the custom tabs feature was the last to be evaluated. Should I change it?

Comment (pocmo) [reply samples/browser/src/main/java/org/mozilla/samples/browser/CustomTabBrowserFragment.kt]: I think it's correct - just looks confusing. I was confused by it too. In this case `super.onBackPressed()` does not call through to the system (like in an activity), but to `BaseBrowserFragment` which will go through the other features first.

Comment (NotWoods) [reply samples/browser/src/main/java/org/mozilla/samples/browser/CustomTabBrowserFragment.kt]: I updated this method with a comment to describe what's happening and clear up the ambiguity.

Comment (jonalmeida) [reply samples/browser/src/main/java/org/mozilla/samples/browser/CustomTabBrowserFragment.kt]: Gotcha, I see it now. Thanks for clarifying!

**Pull request merged**